### ISSUE TITLE
Segmentation fault in backwards step with 3D U-Net on CPU with large input tensor, not OOM

### ISSUE THREAD

**Issue by njneeteson** [pytorch/pytorch#56922]
## 🐛 Bug

I get a segmentation fault during the backward step with a standard 3D U-Net on the CPU with large images (tensor size >= (1,1,360,360,360)).

## To Reproduce

1. Save the below code as a script and run on a machine with at least 120 GB of free RAM and in an environment with just PyTorch (and dependencies) installed.

`

    import torch
    import torch.nn as nn
    import faulthandler

    class Layer(nn.Module):
        def __init__(self,inputs,outputs):

            super(Layer, self).__init__()

            k = 3
            p = (k-1)//2
            s = 1

            self.layer = nn.Sequential(
                nn.Conv3d(inputs,outputs,kernel_size=k,padding=p,stride=s),
                nn.ReLU(inplace=True),
                nn.Conv3d(outputs,outputs,kernel_size=k,padding=p,stride=s),
                nn.ReLU(inplace=True)
            )

        def forward(self,x):
            return self.layer(x)

    class UNet(nn.Module):

        def __init__(self):

            super(UNet, self).__init__()

            self.encoding_layers = nn.ModuleList()
            self.encoding_layers.append(Layer(1,32))
            self.encoding_layers.append(Layer(32,64))
            self.encoding_layers.append(Layer(64,128))
            self.encoding_layers.append(Layer(128,256))

            self.decoding_layers = nn.ModuleList()
            self.decoding_layers.append(Layer(256,128))
            self.decoding_layers.append(Layer(128,64))
            self.decoding_layers.append(Layer(64,32))

            self.downsamples = nn.ModuleList()
            self.downsamples.append(nn.MaxPool3d(2))
            self.downsamples.append(nn.MaxPool3d(2))
            self.downsamples.append(nn.MaxPool3d(2))

            self.upsamples = nn.ModuleList()
            self.upsamples.append(nn.ConvTranspose3d(256,128,kernel_size=2,stride=2))
            self.upsamples.append(nn.ConvTranspose3d(128,64,kernel_size=2,stride=2))
            self.upsamples.append(nn.ConvTranspose3d(64,32,kernel_size=2,stride=2))

            self.map_to_output = nn.Conv3d(32,3,kernel_size=1,stride=1)

        def forward(self,x):

            x1 = self.encoding_layers[0](x)
            x2 = self.encoding_layers[1](self.downsamples[0](x1))
            x3 = self.encoding_layers[2](self.downsamples[1](x2))
            x = self.encoding_layers[3](self.downsamples[2](x3))

            x = self.upsamples[0](x)
            x = self.decoding_layers[0](torch.cat([x3,x],axis=1))

            x = self.upsamples[1](x)
            x = self.decoding_layers[1](torch.cat([x2,x],axis=1))

            x = self.upsamples[2](x)
            x = self.decoding_layers[2](torch.cat([x1,x],axis=1))

            return self.map_to_output(x)

    def main():

        faulthandler.enable()

        model = UNet()

        for L in [32, 64, 128, 256, 400]:

            x = torch.rand(1,1,L,L,L)
            print(f'input shape: {x.shape}')

            out = model(x)
            print(f'output shape: {out.shape}')

            out.sum().backward()
            print(f'Finished for L = {L}')

    if __name__ == '__main__':
        main()
`

Console output:

> input shape: torch.Size([1, 1, 32, 32, 32])
> output shape: torch.Size([1, 3, 32, 32, 32])
> Finished for L = 32
> input shape: torch.Size([1, 1, 64, 64, 64])
> output shape: torch.Size([1, 3, 64, 64, 64])
> Finished for L = 64
> input shape: torch.Size([1, 1, 128, 128, 128])
> output shape: torch.Size([1, 3, 128, 128, 128])
> Finished for L = 128
> input shape: torch.Size([1, 1, 256, 256, 256])
> output shape: torch.Size([1, 3, 256, 256, 256])
> Finished for L = 256
> input shape: torch.Size([1, 1, 400, 400, 400])
> output shape: torch.Size([1, 3, 400, 400, 400])
> Fatal Python error: Segmentation fault
> 
> Thread 0x00007f92361f0700 (most recent call first):
> 
> Current thread 0x00007f9305e22740 (most recent call first):
>   File "/home/fem/anaconda2/envs/bonetorch/lib/python3.7/site-packages/torch/autograd/__init__.py", line 132 in backward
>   File "/home/fem/anaconda2/envs/bonetorch/lib/python3.7/site-packages/torch/tensor.py", line 221 in backward
>   File Segmentation fault (core dumped)

## Expected behavior

I expect the script to finish running without a segmentation fault, as it does for all of the other input tensors before I get to the 400x400x400 dummy image.

## Environment

> Collecting environment information...
> PyTorch version: 1.7.0
> Is debug build: True
> CUDA used to build PyTorch: 10.2
> ROCM used to build PyTorch: N/A
> 
> OS: CentOS Linux release 7.6.1810 (Core)  (x86_64)
> GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-36)
> Clang version: Could not collect
> CMake version: version 3.15.0-rc3
> 
> Python version: 3.7 (64-bit runtime)
> Is CUDA available: True
> CUDA runtime version: 9.0.176
> GPU models and configuration: GPU 0: Quadro P6000
> Nvidia driver version: 430.14
> cuDNN version: Could not collect
> HIP runtime version: N/A
> MIOpen runtime version: N/A
> 
> Versions of relevant libraries:
> [pip3] numpy==1.16.6
> [pip3] torch==1.7.0
> [pip3] torchaudio==0.7.0a0+ac17b64
> [pip3] torchvision==0.8.1
> [conda] blas                      1.0                         mkl  
> [conda] cudatoolkit               10.2.89              hfd86e86_1  
> [conda] mkl                       2020.2                      256  
> [conda] mkl-service               2.3.0            py37he8ac12f_0  
> [conda] mkl_fft                   1.2.0            py37h23d657b_0  
> [conda] mkl_random                1.1.1            py37h0573a6f_0  
> [conda] numpy                     1.16.6           py37hbc911f0_0  
> [conda] numpy-base                1.16.6           py37hde5b4d6_0  
> [conda] pytorch                   1.7.0           py3.7_cuda10.2.89_cudnn7.6.5_0    pytorch
> [conda] torchaudio                0.7.0                      py37    pytorch
> [conda] torchvision               0.8.1                py37_cu102    pytorch

## Additional context

I cannot debug further with GDB because there are debug-infos missing. I can't fix this problem because it is a lab computer that other people rely on and I don't want to break anything, so I cannot trace the segfault back through the stack to the specific problem, unfortunately. I have used a memory profiler to check utilization and ruled out an OOM problem - the forward step uses around 110GB of RAM and the machine I am testing on has over 160GB free RAM. Also the problem occurs in backward after the peak memory utilization has occurred.

cc @ezyang @gchanan @zou3519 @albanD @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @VitalyFedyunin @bdhirsh @jbschlosser @anjali411

**Comment by H-Huang** [comment]
I can also repro this segfault on master
<details>

<summary>Environment</summary>

```
PyTorch version: 1.9.0a0+gita947bfa
Is debug build: True
CUDA used to build PyTorch: 10.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.1 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: 10.0.0-4ubuntu1 
CMake version: version 3.18.2

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.1.105
GPU models and configuration: 
GPU 0: Quadro GP100
GPU 1: Quadro GP100

Nvidia driver version: 450.80.02
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] pytorch-sphinx-theme==0.0.24
[pip3] torch==1.9.0a0+git963f005
[pip3] torchaudio==0.8.1
[pip3] torchtext==0.9.1
[pip3] torchvision==0.9.1
[conda] blas                      1.0                         mkl  
[conda] magma-cuda102             2.5.2                         1    pytorch
[conda] mkl                       2020.2                      256  
[conda] mkl-include               2020.2                      256  
[conda] mkl-service               2.3.0            py38he904b0f_0  
[conda] mkl_fft                   1.2.0            py38h23d657b_0  
[conda] mkl_random                1.1.1            py38h0573a6f_0  
[conda] numpy                     1.19.2           py38h54aff64_0  
[conda] numpy-base                1.19.2           py38hfa32c7d_0  
[conda] torch                     1.9.0a0+git963f005           dev_0    <develop>
```

</details>

**Comment by pzharrington** [comment]
Just chiming in here: I can get a segfault even with a simple Conv3d if I use a sufficiently large input tensor (I ended up at this issue since my U-Net is also hitting segfaults during inference on large inputs on CPU).

A simple reproducer is
```
import torch
import torch.nn as nn
import numpy as np
size=512
sample = torch.from_numpy(np.ones((1, 4, size,size,size), dtype=np.single))
conv = nn.Conv3d(4, 64, 3, stride=1, padding=1)
x = conv(sample)
```

For me, the above segfaults for `size=512` and above, and it is not due to running out of CPU memory (running on a system with 512GB of RAM).

**Comment by jbschlosser** [comment]
I can reproduce this with @pzharrington's script. Disabling mkl-dnn fixes the issue for me:
```python
with torch.backends.mkldnn.flags(enabled=False):
    x = conv(sample)
```

cc @XiaobingSuper

**Comment by XiaobingSuper** [comment]
Thanks, @jbschlosser, I can reproduce this issue, and we will fix it as soon as possible.

**Comment by XiaobingSuper** [comment]
This issue will be fixed by OneDNN 2.7, the related fixed commit is https://github.com/oneapi-src/oneDNN/commit/3a71229914c4cbd76de672e9206f9b0b37304480
